kubelet tracing instrumentation

sallyom opened the pull request:
Signed-off-by: Sally O'Malley <somalley@redhat.com>

#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

[Issue](https://github.com/kubernetes/enhancements/issues/2831)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
No
```
[KEP](https://github.com/kubernetes/enhancements/pull/2832)

sallyom commented:
![crio-kubelet-exec](https://user-images.githubusercontent.com/11166065/133914508-f625de56-9207-43ba-ad2f-59ae58479d31.png)
![trace-ui-etcd-apiserver-crio-kubelet](https://user-images.githubusercontent.com/11166065/133914512-a3232810-7b7e-41e7-97d2-f7540e919106.png)

sallyom commented:
sifting through code, placing instrumentation - it's not all in the ideal places yet - I see kubelet<->crio and etcd<->apiserver, but looking at where/how/if can propagate kubelet <-> apiserver 

/hold

ehashman commented:
/label api-review

caesarxuchao commented:
/remove-sig api-machinery

sallyom committed:
kubelet tracing

Signed-off-by: Sally O'Malley <somalley@redhat.com>
Co-authored-by: David Ashpole <dashpole@google.com>

sallyom commented:
The component-base/traces/config.go,config_test.go are mostly taken from @dashpole 's addition to `apiserver/pkg/tracing/config.go,config_test.go`. I've added him as a co-author because of this, and because this addition closely follows the tracing addition to apiserver, with his added suggestion of adding a tracing-config to component-base. I'll implement changes from reviews, but this is based off of his work with apiserver.

sallyom commented:
/label api-review

/hold cancel

fedebongio commented:
/remove-sig api-machinery

dashpole commented:
/triage accepted

dashpole commented:
/assign

dims commented:
/retest

@sallyom the verify failures are real ones for sure. kicking all the failed tests to run again just in case

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -951,6 +954,22 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) {
+		opts := []otlpgrpc.Option{}
+		if s.KubeletConfiguration.Tracing.Endpoint == nil {
```
I think this should be `!=`

sallyom replied:
whoops fixed

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -108,6 +110,7 @@ import (
 const (
 	// Kubelet component name
 	componentKubelet = "kubelet"
+	kubeletClientService = "kubelet-client"
```
This should probably just be "kubelet" for everything in this PR (clients or servers)

sallyom replied:
updated (removed, pass componentKubelet instead)

dashpole commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -73,7 +75,11 @@ func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration) (
 
-	conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure(), grpc.WithContextDialer(dialer), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)))
+	var conn *grpc.ClientConn
+	conn, err = grpc.DialContext(ctx, addr, grpc.WithInsecure(), grpc.WithContextDialer(dialer), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)))
```
nit: probably better to define opts, append, and consolidate the DialContext calls:
```
opts := grpc.WithInsecure(), grpc.WithContextDialer(dialer), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize))
if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) {
  opts = append(opts, grpc.WithUnaryInterceptor(otelgrpc.UnaryClientInterceptor()), grpc.WithStreamInterceptor(otelgrpc.StreamClientInterceptor()))
}
conn, err = grpc.DialContext(...
```

sallyom replied:
right, updated, thanks

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -130,6 +135,8 @@ type filteringContainer struct {
 
 func (a *filteringContainer) Handle(path string, handler http.Handler) {
+	// no tracer provider then noop
+	handler = traces.WithTracing(handler, otel.GetTracerProvider(), "kubelet-container")
```
even though it is a no-op, we should still feature gate this.  We should also try and plumb the tracerprovider through to here, rather than use the global one.

sallyom replied:
embedded TracerProvider in filteringContainer struct

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -240,6 +248,20 @@ func NewServer(
+			opts = append(opts, otlpgrpc.WithEndpoint(*kubeCfg.Tracing.Endpoint))
+		}
+		opts = append(opts, otlpgrpc.WithInsecure())
+		tp, err := traces.InitTracing(context.Background(), host.GetHostname(), kubeletServiceName, *kubeCfg.Tracing.SamplingRatePerMillion, opts)
```
Ideally, we should only construct the tracer provider once in the kubelet, and should plumb it through to wherever it is needed (grpc/http handlers)

dashpole commented on staging/src/k8s.io/component-base/config/v1alpha1/defaults.go:
```diff
@@ -126,3 +126,18 @@ func RecommendedLoggingConfiguration(obj *LoggingConfiguration) {
+// run it in your wrapper struct of this type in its `SetDefaults_` method.
+func RecommendedTracingConfiguration(obj *TracingConfiguration) {
+	if obj.Endpoint == nil {
+		obj.Endpoint = utilpointer.String("localhost:4317")
```
Can we just rely on the fact that otel's default is already :4317, and not set our own value here?

sallyom replied:
yes, removed this and the call to it
from godoc
`If unset, it will instead try to use connect to DefaultCollectorHost:DefaultCollectorPort` localhost:4317

dashpole commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,33 +26,64 @@ import (
+func WithTracing(handler http.Handler, tp trace.TracerProvider, serviceName string) http.Handler {
+	opts := []otelhttp.Option{
+		otelhttp.WithPropagators(Propagators()),
+		otelhttp.WithPublicEndpoint(),
```
Note that WithPublicEndpoint should cause incoming context to be ignored (e.g. would prevent apiserver -> kubelet traces from being linked).  If the handler is on the authenticated port, and comes after auth is handled in the server stack, we don't need this.

sallyom replied:
I would def not have seen that, thanks, removed

dashpole commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,33 +26,64 @@ import (
+	samplingRate *int32,
+	opts []otlpgrpc.Option,
+) (trace.TracerProvider, error) {
+	resourceOpts := []resource.Option{
```
I realize the resource code is duplicated in components, but i'd still prefer passing in resourceOpts as a parameter so that components can easily add resource attributes if they want in the future.

sallyom replied:
makes sense, ok

dashpole commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,33 +26,64 @@ import (
+func InitTracing(ctx context.Context,
+	traceID string,
+	serviceName string,
+	samplingRate *int32,
```
Could this take a componentbase.TracingConfiguration?  That would allow extending the config without modifying the function signature in the future.

dashpole commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -73,7 +75,12 @@ func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration) (
+	dialOpts := []grpc.DialOption{}
+	dialOpts = append(dialOpts, grpc.WithInsecure(), grpc.WithContextDialer(dialer), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)))
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) {
+		dialOpts = append(dialOpts, grpc.WithUnaryInterceptor(otelgrpc.UnaryClientInterceptor()), grpc.WithStreamInterceptor(otelgrpc.StreamClientInterceptor()))
```
We should plumb the tracer providers through to here.

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -240,11 +253,27 @@ func NewServer(
+			klog.ErrorS(err, "Failed to configure kubelet traces")
+			os.Exit(1)
+		}
+		otel.SetTracerProvider(tp)
```
If possible, we should avoid using the global tracer provider and propagators.  For each client/server, we should pass propagators and tracer provider explicitly.

sallyom replied:
done, removed this.

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -240,11 +253,27 @@ func NewServer(
+		}
+		opts = append(opts, otlpgrpc.WithInsecure())
+		var err error
+		tp, err = traces.InitTracing(context.Background(), host.GetHostname(), kubeletServiceName, kubeCfg.Tracing.SamplingRatePerMillion, opts)
```
If possible, we should re-use the tracer-provider created in cmd/kubelet/app.  kubedeps is probably the right place to store it.

sallyom replied:
ok, got it. I added to kubeDeps

ritazh commented:
/remove-sig auth Feel free to add us back if needed.

sallyom commented on pkg/kubelet/server/server.go:
```diff
@@ -240,11 +253,27 @@ func NewServer(
+	var tp oteltrace.TracerProvider
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) {
+		opts := []otlpgrpc.Option{}
+		if kubeCfg.Tracing.Endpoint == nil {
```
weird, i did fix this :shrug:

sallyom replied:
oh i did that twice, that's why - removed anyways fixed

sallyom commented:
screenshots with kubelet, cri-o, etcd, and apiserver traces can be viewed [here](https://github.com/sallyom/otel-kubeadm#readme)

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -400,6 +407,33 @@ func UnsecuredDependencies(s *options.KubeletServer, featureGate featuregate.Fea
+		    return nil, err
+	    }
+	    opts := []otlpgrpc.Option{}
+		tc := componentbaseconfigv1alpha1.TracingConfiguration{
```
isn't `tc` the same as `s.KubeletConfiguration.Tracing`?

sallyom replied:
yes, i must've been tired

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -173,9 +183,14 @@ func ListenAndServeKubeletServer(
 
 // ListenAndServeKubeletReadOnlyServer initializes a server to respond to HTTP network requests on the Kubelet.
-func ListenAndServeKubeletReadOnlyServer(host HostInterface, resourceAnalyzer stats.ResourceAnalyzer, address net.IP, port uint) {
+func ListenAndServeKubeletReadOnlyServer(
```
For now, it may be best to omit tracing on the readonly server, since it is not authenticated.  Ideally, this server would use the WithPublicEndpoint option we removed elsewhere, and the authenticated server would not use the WithPublicEndpoint option.  That way, an unauthenticated user cannot control sampling decisions in the kubelet.

dashpole replied:
Opened https://github.com/kubernetes/kubernetes/issues/109829 to track.

liggitt replied:
sounds good... add a TODO with the reference to that issue here where we construct NewNoopTracerProvider?

dashpole commented on staging/src/k8s.io/apiserver/pkg/server/options/tracing.go:
```diff
@@ -70,15 +70,21 @@ func (o *TracingOptions) ApplyTo(es *egressselector.EgressSelector, c *server.Co
+	}
 	opts := []otlpgrpc.Option{}
 	if npConfig.Endpoint != nil {
+		tc.Endpoint = npConfig.Endpoint
```
I think we already set this above.

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -400,6 +407,33 @@ func UnsecuredDependencies(s *options.KubeletServer, featureGate featuregate.Fea
+			semconv.ServiceInstanceIDKey.String(componentKubelet),
+		),
+	    }
+		if s.KubeletConfiguration.Tracing.Endpoint != nil {
```
can we add the otelgrpc.WithEndpoint option from within `traces.InitTracing`, and also add `otelgrpc.WithInsecure` there?

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -400,6 +407,33 @@ func UnsecuredDependencies(s *options.KubeletServer, featureGate featuregate.Fea
+		if err != nil {
+			return nil, err
+		}
+		otel.SetTextMapPropagator(traces.Propagators())
```
Setting the global textmappropagator shouldn't be necessary.  On the http and grpc handlers, we just need to pass a WithPropagators (or something) option explicitly.

dashpole reviewed (commented):
This is looking really nice overall

helayoty commented:
@sallyom 👋  I'm in the release 1.24 bug triage shadow. Does this issue still target release 1.24? If yes, would you please update the priority field for us to track it better?

sallyom commented:
> @sallyom wave I'm in the release 1.24 bug triage shadow. Does this issue still target release 1.24? If yes, would you please update the priority field for us to track it better?

Yes, enhancement is expected to make it in for 1.24, and also this. I'm not sure how to set the priority field, though. :eyes:

helayoty commented:
> > @sallyom wave I'm in the release 1.24 bug triage shadow. Does this issue still target release 1.24? If yes, would you please update the priority field for us to track it better?
> 
> Yes, enhancement is expected to make it in for 1.24, and also this. I'm not sure how to set the priority field, though. 👀

You can add the priority by adding a tag(/tag-name) to this PR to one of [these](https://github.com/kubernetes/kubernetes/labels?q=priority) tags.

sallyom commented:
/priority/important-soon

pacoxu commented on go.mod:
```diff
@@ -81,6 +81,9 @@ require (
 	go.etcd.io/etcd/client/v3 v3.5.0
+	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc v0.20.0
+	go.opentelemetry.io/otel v0.20.0
+	go.opentelemetry.io/otel/exporters/otlp v0.20.0
```
Is there some special reason that we don't use v1.4.0?

https://pkg.go.dev/go.opentelemetry.io/otel@v1.4.0?tab=versions

- v1.* is for v1
- v0.20.0 is for v0

dashpole replied:
There are kubernetes dependencies (etcd, containerd IIRC) which use older versions of otel.  We are stuck on those until we can upgrade dependencies.

helayoty commented:
@sallyom @ehashman 👋  the release 1.24 bug triage shadow. Does this PR still target release 1.24?

sallyom commented:
> @sallyom @ehashman wave the release 1.24 bug triage shadow. Does this PR still target release 1.24? If yes, could you please update the priority?

yes I've updated to urgent to get this reviewed in next few days, :crossed_fingers:

sallyom commented:
/priority/critical-urgent

sallyom commented:
@dashpole ptal, goal is to get this in by 1.24 code-freeze, any chance?

sallyom commented:
/label priority/critical-urgent

dashpole reviewed (commented):
Some follow-ups:
* We should use a NoopTracerProvider instead of nil when tracing is disabled.  That way, we don't need to check for nil everywhere anymore.
* We should add an integration test similar to the apiserver one.
* Set endpoint and insecure in InitTracing()

dashpole commented:
/retest

helayoty commented:
/priority critical-urgent

dashpole commented:
/lgtm
/approve

ehashman commented:
/remove-priority critical-urgent
/priority important-soon

please only tag urgent/release-blocking bug fixes with critical-urgent, thanks

ehashman commented on cmd/kubelet/app/server.go:
```diff
@@ -367,6 +373,32 @@ func UnsecuredDependencies(s *options.KubeletServer, featureGate featuregate.Fea
 	}
+	var tp oteltrace.TracerProvider
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) {
+		hostname, err := nodeutil.GetHostname(s.HostnameOverride)
+		if err != nil {
+			return nil, err
+		}
```
If we do this, I think this means if the hostname can't be found we fail to return UnsecuredDependencies and it would fail kubelet startup... do we want this to fail a bit more gracefully as this is an alpha feature? Log an error and leave `tp` as nil?

ehashman commented on cmd/kubelet/app/server.go:
```diff
@@ -367,6 +373,32 @@ func UnsecuredDependencies(s *options.KubeletServer, featureGate featuregate.Fea
+		}
+		tp, err = traces.InitTracing(context.Background(), *tc, opts, resourceOpts)
+		if err != nil {
+			return nil, err
```
Same comment as above re: returning an errors in this block. I wonder if it would make sense to factor this all out into a helper function, as this function previously didn't do very much. The logic for the other plugins isn't included here.

sallyom replied:
agreed, thanks, updated

ehashman commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -440,10 +440,13 @@ type KubeletConfiguration struct {
 	// registerNode enables automatic registration with the apiserver.
 	// +optional
 	RegisterNode bool
+	// Tracing specifies the options of tracing.
```
Could we add a bit more information here? A user might not know what tracing is. Particularly because we default this, I'd expect the API docs to tell us what the defaults are.

sallyom replied:
done

ehashman commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -213,6 +214,12 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 
+	if localFeatureGate.Enabled(features.KubeletTracing) {
+		if errs := traces.ValidateTracingConfiguration(kc.Tracing, field.NewPath("tracing")); len(errs) > 0 {
+			allErrors = append(allErrors, errs.ToAggregate().Errors()...)
```
Yes, please add a test to https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/validation/validation_test.go

ehashman replied:
resolved

ehashman commented on pkg/kubelet/server/server.go:
```diff
@@ -239,12 +253,17 @@ func NewServer(
 	auth AuthInterface,
+	tp *oteltrace.TracerProvider,
 	kubeCfg *kubeletconfiginternal.KubeletConfiguration) Server {
+	rc := &filteringContainer{Container: restful.NewContainer()}
+	if tp != nil {
+		rc = &filteringContainer{Container: restful.NewContainer(), TracerProvider: *tp}
+	}
```
Should this be feature gated?

sallyom replied:
yes, i think so, added

ehashman commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go:
```diff
@@ -150,6 +150,7 @@ type TLSConfig struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // TracingConfiguration provides versioned configuration for tracing clients.
+// This type will go away in favor of component-base/config TracingConfiguration
```
Is this intended to be a deprecation notice?

```suggestion
// This type is deprecated in favor of component-base/config TracingConfiguration
```

ehashman commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -151,6 +151,7 @@ type TLSConfig struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // TracingConfiguration provides versioned configuration for tracing clients.
+// This type will go away in favor of component-base/config TracingConfiguration
```
Same comment as above

sallyom replied:
ok, updated here, too

ehashman commented on staging/src/k8s.io/apiserver/pkg/server/options/tracing.go:
```diff
@@ -19,24 +19,37 @@ package options
 	"k8s.io/apiserver/pkg/server"
 	"k8s.io/apiserver/pkg/server/egressselector"
-	"k8s.io/apiserver/pkg/tracing"
+	v1alpha1componentbaseconfig "k8s.io/component-base/config/v1alpha1"
```
Since the old types aren't going away, do we need to do this now or can we save the cleanup for later?

sallyom replied:
this PR updates the handling of TracingConfiguration to the component-base, but accepts the apiserver.TracingConfiguration read in from the config file. As soon as the internal tracing configuration is decoded from the config file, it is converted into the new component-base/config struct.

ehashman commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,16 @@ func (l *VerbosityLevel) Set(value string) error {
 	return "Level"
 }
+
+// TracingConfiguration provides versioned configuration for tracing clients.
```
```suggestion
// TracingConfiguration provides versioned configuration for OpenTelemetry tracing clients.
```

It's otel only right?

sallyom replied:
yes, updated with latest push

ehashman commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,16 @@ func (l *VerbosityLevel) Set(value string) error {
+
+// TracingConfiguration provides versioned configuration for tracing clients.
+type TracingConfiguration struct {
+	// Endpoint of the collector this component will report to.
```
```suggestion
	// Endpoint of the collector this component will report traces to.
```

sallyom replied:
included in pushed updates

ehashman commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,16 @@ func (l *VerbosityLevel) Set(value string) error {
+type TracingConfiguration struct {
+	// Endpoint of the collector this component will report to.
+	// The syntax is defined in https://github.com/grpc/grpc/blob/master/doc/naming.md.
+	// Recommended default is the otlp grpc default, localhost:4317
```
"Recommended default" is a bit confusing... is this the recommended value, or the default? What happens if this is unset?

sallyom replied:
updated, if unset and enabled, it is the default  "localhost:4317"

ehashman commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -778,6 +778,10 @@ type KubeletConfiguration struct {
 	// Default: true
 	// +optional
 	RegisterNode *bool `json:"registerNode,omitempty"`
+	// Tracing specifies the options of tracing.
```
Same comments as on the internal type -- can we get a bit more info here, as well as what the defaults are?

ehashman commented on cmd/kubelet/app/server.go:
```diff
@@ -367,6 +373,32 @@ func UnsecuredDependencies(s *options.KubeletServer, featureGate featuregate.Fea
+		}
+		opts = append(opts, otlpgrpc.WithInsecure())
+		tc := &configv1alpha1.TracingConfiguration{}
+		if err := configv1alpha1.Convert_config_TracingConfiguration_To_v1alpha1_TracingConfiguration(&s.KubeletConfiguration.Tracing, tc, nil); err != nil {
```
Is it necessary to do this conversion explicitly? I don't think I've seen anything like this in a config API...

/cc @liggitt

sallyom replied:
updated to component-base/config/v1alpha1 TracingConfiguration everywhere

sallyom replied:
i think i have it sorted out now, ptal

sallyom replied:
ptal again, thank you

ehashman reviewed (commented):
API review + a pass for node as well.

Note that I don't see any e2es included here, KEP alpha has a graduation goal of "Integration testing of tracing". Was that manually tested?

Please also update the release note. This shouldn't need an action required afaict.

ehashman commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -778,6 +778,15 @@ type KubeletConfiguration struct {
+	// See http://kep.k8s.io/2832 for more details.
+	// +featureGate=KubeletTracing
+	// +optional
+	Tracing componentbaseconfigv1alpha1.TracingConfiguration `json:"tracing,omitempty"`
```
Note: this matches `componentbaseconfigv1alpha1.LoggingConfiguration` above so I don't think this is making anything worse here.

ehashman commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -67,7 +71,7 @@ const (
 
 // NewRemoteRuntimeService creates a new internalapi.RuntimeService.
-func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration) (internalapi.RuntimeService, error) {
+func NewRemoteRuntimeService(endpoint string, tp *trace.TracerProvider, connectionTimeout time.Duration) (internalapi.RuntimeService, error) {
```
It is a little weird to be inserting tp as the middle argument, can you move that to the end?

sallyom replied:
updated, moved to end

ehashman commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -441,10 +441,18 @@ type KubeletConfiguration struct {
 	// +optional
 	RegisterNode bool
+	// Tracing specifies the versioned configuration for OpenTelemetry tracing clients.
+	// Specifies the endpoint of the collector that kubelet will traces report to.
+	// The default endpoint is the otlp grpc default, localhost:4317
+	// Specifies the number of samples to collect per million spans.
+	// Sampling rate defaults to 0.
```
The current defaults are nil, and these are the recommended settings. I suggest we remove these details here.

```suggestion
```

ehashman commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,16 @@ func (l *VerbosityLevel) Set(value string) error {
+type TracingConfiguration struct {
+	// Endpoint of the collector this component will report traces to.
+	// The syntax is defined in https://github.com/grpc/grpc/blob/master/doc/naming.md.
+	// Recommended default is the otlp grpc default, localhost:4317
```
```suggestion
	// Recommended setting is the otlp grpc default, localhost:4317
```

ehashman commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,16 @@ func (l *VerbosityLevel) Set(value string) error {
+	Endpoint *string
+
+	// SamplingRatePerMillion is the number of samples to collect per million spans.
+	// Defaults to 0.
```
```suggestion
	// +featureGate=KubeletTracing
	// +optional
```

This defaults to nil/unset so I don't think this is accurate?

ehashman commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,16 @@ func (l *VerbosityLevel) Set(value string) error {
+	// Endpoint of the collector this component will report traces to.
+	// The syntax is defined in https://github.com/grpc/grpc/blob/master/doc/naming.md.
+	// Recommended default is the otlp grpc default, localhost:4317
+	// The connection is insecure, and does not currently support TLS.
```
Don't forget to add the feature gate and optional info.

```suggestion
	// The connection is insecure, and does not currently support TLS.
	// +featureGate=KubeletTracing
	// +optional
```

ehashman commented on staging/src/k8s.io/component-base/config/v1alpha1/types.go:
```diff
@@ -141,3 +141,18 @@ type VModuleItem struct {
+	// Endpoint of the collector this component will report traces to.
+	// The syntax is defined in https://github.com/grpc/grpc/blob/master/doc/naming.md.
+	// Default is the otlp grpc default, localhost:4317
+	// The connection is insecure, and does not currently support TLS.
```
See comments on the internal type above.

sallyom replied:
updated, ptal, it's now the same in both config & config/v1alpha1 as below:
```go
// TracingConfiguration provides versioned configuration for OpenTelemetry tracing clients.
type TracingConfiguration struct {
    // Endpoint of the collector this component will report traces to.
    // The connection is insecure, and does not currently support TLS.
    // Recommended is unset, and endpoint is the otlp grpc default, localhost:4317.
    // +featureGate=KubeletTracing
    // +optional
    Endpoint *string

    // SamplingRatePerMillion is the number of samples to collect per million spans.
    // Recommended is unset. If unset, sampler respects its parent span's sampling
    // rate, but otherwise never samples.
    // +featureGate=KubeletTracing
    // +optional
    SamplingRatePerMillion *int32
}

```

ehashman commented on staging/src/k8s.io/component-base/config/v1alpha1/types.go:
```diff
@@ -141,3 +141,18 @@ type VModuleItem struct {
+
+	// +optional
+	// SamplingRatePerMillion is the number of samples to collect per million spans.
+	// Defaults to 0.
```
See comments on the internal type above.

sallyom replied:
removed the `Default` line and modified

ehashman reviewed (commented):
This looks close to me. Few more things before I think this is ready for API approval.

dims commented:
@sallyom needs rebase please!

sallyom commented:
@ehashman please see the last 2 commits (and your committed suggestions) I hope it's now ready for API final review, thanks!

ehashman commented:
Hi @sallyom, before I do a final pass on this, can you please address the test failures?

We have a gofmt issue as well as a failing unit test:

```
{Failed  === RUN   TestComponentConfigSetup
    register_test.go:34: [external types should have json tags. schema.GroupVersionKind{Group:"kubelet.config.k8s.io", Version:"v1beta1", Kind:"KubeletConfiguration"} tags on field Endpoint are: .
        Type parents:
        v1beta1.KubeletConfiguration
         v1alpha1.TracingConfiguration
        , external types should have json tags. schema.GroupVersionKind{Group:"kubelet.config.k8s.io", Version:"v1beta1", Kind:"KubeletConfiguration"} tags on field SamplingRatePerMillion are: .
        Type parents:
        v1beta1.KubeletConfiguration
         v1alpha1.TracingConfiguration
        ]
--- FAIL: TestComponentConfigSetup (0.00s)
```

sallyom commented:
> Hi @sallyom, before I do a final pass on this, can you please address the test failures?
> 
> We have a gofmt issue as well as a failing unit test:

@ehashman should be all set now, sorry about that, thanks

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -441,10 +441,14 @@ type KubeletConfiguration struct {
+	// See http://kep.k8s.io/2832 for more details.
+	// +featureGate=KubeletTracing
+	// +optional
+	Tracing componentbaseconfig.TracingConfiguration
```
make this a pointer

sallyom replied:
updated

liggitt commented on pkg/kubelet/apis/config/scheme/testdata/KubeletConfiguration/roundtrip/default/v1beta1.yaml:
```diff
@@ -85,5 +85,6 @@ streamingConnectionIdleTimeout: 4h0m0s
 syncFrequency: 1m0s
 topologyManagerPolicy: none
 topologyManagerScope: container
+tracing: {}
```
this should be nil / omitted by default

sallyom replied:
removed

liggitt commented on pkg/kubelet/apis/config/scheme/testdata/KubeletConfiguration/after/v1beta1.yaml:
```diff
@@ -85,5 +85,6 @@ streamingConnectionIdleTimeout: 4h0m0s
 syncFrequency: 1m0s
 topologyManagerPolicy: none
 topologyManagerScope: container
+tracing: {}
```
this should be nil / omitted by default after round-trip

sallyom replied:
removed

liggitt commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -241,6 +242,12 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 		allErrors = append(allErrors, errs.ToAggregate().Errors()...)
 	}
 
+	if localFeatureGate.Enabled(features.KubeletTracing) {
```
reject a config file that tries to enable tracing if the feature is off

```go
if kc.Tracing != nil {
  if !localFeatureGate.Enabled(features.KubeletTracing) {
    ... error, cannot use tracing if KubeletTracing feature is not enabled
  } else {
    allErrors = append(allErrors, traces.ValidateTracingConfiguration(kc.Tracing, field.NewPath("tracing"))...)
  }
}
```

sallyom replied:
updated, thanks

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/tracing.go:
```diff
@@ -65,20 +79,24 @@ func (o *TracingOptions) ApplyTo(es *egressselector.EgressSelector, c *server.Co
 		return nil
 	}
 
```
@dashpole - this is pre-existing, but I'm wondering if this is the point at which we should be catching an attempt to enable tracing in the API server if the feature is not enabled

That would mean if o.ConfigFile != "", and we got here, and the feature wasn't enabled, we'd return an error.

That would also mean calls to `TracingOptions#ApplyTo` should not be feature-gated

dashpole replied:
I haven't seen that pattern (validating that feature-gates are enabled if the corresponding config is enabled), but it makes sense to me.

liggitt replied:
otherwise, they can set a flag or config field, the binary silently ignores it, and then their identical invocation breaks on upgrade then feature leaps into action

sallyom replied:
not sure you wanted this change in this PR, but added it in the last commit
PTAL, thanks

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -126,10 +128,16 @@ type containerInterface interface {
 // so we can ensure restful.FilterFunctions are used for all handlers
 type filteringContainer struct {
 	*restful.Container
+	oteltrace.TracerProvider
```
We should be using the NoopTracerProvider instead of nil everywhere, so we don't have to handle nil checking.  I was planning to change that as a follow-up across both apiserver tracing and kubelet tracing.

sallyom replied:
I updated to use NoopTracerProvider  in the last pushed commit, PTAL, thanks

sallyom commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,33 +26,64 @@ import (
+		return nil, err
 	}
 
+	// Only emit spans when the kubelet sends a request with a sampled trace
```
missed that, thanks, updated

sallyom commented on pkg/kubelet/server/server.go:
```diff
@@ -239,12 +253,17 @@ func NewServer(
+	tp *oteltrace.TracerProvider,
 	kubeCfg *kubeletconfiginternal.KubeletConfiguration) Server {
+	rc := &filteringContainer{Container: restful.NewContainer()}
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) && tp != nil {
```
updated - still have to check for non-nil, since passing tp not *tp to filteringContainer - this ok?

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -143,12 +150,13 @@ func ListenAndServeKubeletServer(
 	tlsOptions *TLSOptions,
-	auth AuthInterface) {
+	auth AuthInterface,
+	tp *oteltrace.TracerProvider) {
```
```suggestion
	tp oteltrace.TracerProvider) {
```

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -239,12 +252,18 @@ func NewServer(
 	host HostInterface,
 	resourceAnalyzer stats.ResourceAnalyzer,
 	auth AuthInterface,
+	tp *oteltrace.TracerProvider,
```
```suggestion
	tp oteltrace.TracerProvider,
```

dashpole commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,49 +26,72 @@ import (
 //     config.Wrap(traces.WrapperFor(&tp))
 //     kubeclient, _ := clientset.NewForConfig(config)
-func WrapperFor(tp *trace.TracerProvider) transport.WrapperFunc {
+func WrapperFor(tp *oteltrace.TracerProvider) transport.WrapperFunc {
```
```suggestion
func WrapperFor(tp oteltrace.TracerProvider) transport.WrapperFunc {
```

dashpole commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,49 +26,72 @@ import (
 	return func(rt http.RoundTripper) http.RoundTripper {
 		opts := []otelhttp.Option{
 			otelhttp.WithPropagators(Propagators()),
+			otelhttp.WithTracerProvider(*tp),
```
```suggestion
			otelhttp.WithTracerProvider(tp),
```

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -786,7 +796,7 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
 // buildKubeletClientConfig constructs the appropriate client config for the kubelet depending on whether
 // bootstrapping is enabled or client certificate rotation is enabled.
-func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nodeName types.NodeName) (*restclient.Config, func(), error) {
+func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, tp *oteltrace.TracerProvider, nodeName types.NodeName) (*restclient.Config, func(), error) {
```
```suggestion
func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, tp oteltrace.TracerProvider, nodeName types.NodeName) (*restclient.Config, func(), error) {
```

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -1291,3 +1304,32 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
 	return rl, nil
 }
+
+func configureTracerProvider(s *options.KubeletServer) *oteltrace.TracerProvider {
```
```suggestion
func configureTracerProvider(s *options.KubeletServer) oteltrace.TracerProvider {
```

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -1291,3 +1304,32 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
+
+func configureTracerProvider(s *options.KubeletServer) *oteltrace.TracerProvider {
+	if s.KubeletConfiguration.Tracing == nil {
+		tp := oteltrace.NewNoopTracerProvider()
+		return &tp
```
```suggestion
		return oteltrace.NewNoopTracerProvider()
```

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -1291,3 +1304,32 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
+	hostname, err := nodeutil.GetHostname(s.HostnameOverride)
+	if err != nil {
+		klog.V(2).InfoS("Warning: Could not configure tracer provider", "error", err)
+		return nil
```
```suggestion
		return oteltrace.NewNoopTracerProvider()
```

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -1291,3 +1304,32 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
+	tp, err := traces.InitTracing(context.Background(), s.KubeletConfiguration.Tracing, opts, resourceOpts)
+	if err != nil {
+		klog.V(2).InfoS("Warning: Could not configure tracer provider", "error", err)
+		return nil
```
```suggestion
		return oteltrace.NewNoopTracerProvider()
```

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -1291,3 +1304,32 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
+		klog.V(2).InfoS("Warning: Could not configure tracer provider", "error", err)
+		return nil
+	}
+	return &tp
```
```suggestion
	return tp
```

dashpole commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -67,7 +71,7 @@ const (
 
 // NewRemoteRuntimeService creates a new internalapi.RuntimeService.
-func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration) (internalapi.RuntimeService, error) {
+func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration, tp *trace.TracerProvider) (internalapi.RuntimeService, error) {
```
```suggestion
func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration, tp trace.TracerProvider) (internalapi.RuntimeService, error) {
```

dashpole commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -76,7 +80,22 @@ func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration) (
+	dialOpts := []grpc.DialOption{}
+	dialOpts = append(dialOpts, grpc.WithInsecure(), grpc.WithContextDialer(dialer), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)))
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) {
+		tracingOpts := []otelgrpc.Option{
+			otelgrpc.WithPropagators(traces.Propagators()),
+		}
+		if tp != nil {
+			tracingOpts = append(tracingOpts, otelgrpc.WithTracerProvider(*tp))
+		}
```
```suggestion
		tracingOpts := []otelgrpc.Option{
			otelgrpc.WithPropagators(traces.Propagators()),
			otelgrpc.WithTracerProvider(tp),
		}
```

dashpole commented on pkg/kubelet/cri/remote/remote_runtime_test.go:
```diff
@@ -47,7 +47,7 @@ func createAndStartFakeRemoteRuntime(t *testing.T) (*fakeremote.RemoteRuntime, s
 
 func createRemoteRuntimeService(endpoint string, t *testing.T) internalapi.RuntimeService {
-	runtimeService, err := NewRemoteRuntimeService(endpoint, defaultConnectionTimeout)
+	runtimeService, err := NewRemoteRuntimeService(endpoint, defaultConnectionTimeout, nil)
```
Needs NewNoopTracerProvider instead of nil.

sallyom replied:
updated

dashpole commented on pkg/kubelet/kubelet.go:
```diff
@@ -205,7 +206,7 @@ type Bootstrap interface {
 	BirthCry()
 	StartGarbageCollection()
-	ListenAndServe(kubeCfg *kubeletconfiginternal.KubeletConfiguration, tlsOptions *server.TLSOptions, auth server.AuthInterface)
+	ListenAndServe(kubeCfg *kubeletconfiginternal.KubeletConfiguration, tlsOptions *server.TLSOptions, auth server.AuthInterface, tp *trace.TracerProvider)
```
```suggestion
	ListenAndServe(kubeCfg *kubeletconfiginternal.KubeletConfiguration, tlsOptions *server.TLSOptions, auth server.AuthInterface, tp trace.TracerProvider)
```

dashpole commented on pkg/kubelet/kubelet.go:
```diff
@@ -235,6 +236,7 @@ type Dependencies struct {
 	ProbeManager         prober.Manager
 	Recorder             record.EventRecorder
 	Subpather            subpath.Interface
+	TracerProvider       *trace.TracerProvider
```
```suggestion
	TracerProvider       trace.TracerProvider
```

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -173,9 +182,13 @@ func ListenAndServeKubeletServer(
+	port uint) {
 	klog.InfoS("Starting to listen read-only", "address", address, "port", port)
-	s := NewServer(host, resourceAnalyzer, nil, nil)
+	s := NewServer(host, resourceAnalyzer, nil, nil, nil)
```
Needs NewNoopTracerProvider instead of nil.

sallyom replied:
updated

dashpole commented on pkg/kubelet/server/server.go:
```diff
@@ -239,12 +252,18 @@ func NewServer(
 	auth AuthInterface,
+	tp *oteltrace.TracerProvider,
 	kubeCfg *kubeletconfiginternal.KubeletConfiguration) Server {
+	rc := &filteringContainer{Container: restful.NewContainer()}
+	if tp != nil {
+		rc = &filteringContainer{Container: restful.NewContainer(), TracerProvider: *tp}
+	}
```
```suggestion
	rc := &filteringContainer{Container: restful.NewContainer(), TracerProvider: tp}
```

dashpole commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/traces.go:
```diff
@@ -30,11 +30,9 @@ func WithTracing(handler http.Handler, tp *trace.TracerProvider) http.Handler {
 	opts := []otelhttp.Option{
 		otelhttp.WithPropagators(traces.Propagators()),
 		otelhttp.WithPublicEndpoint(),
+		otelhttp.WithTracerProvider(*tp),
```
```suggestion
		otelhttp.WithTracerProvider(tp),
```

dashpole commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/traces.go:
```diff
@@ -30,11 +30,9 @@ func WithTracing(handler http.Handler, tp *trace.TracerProvider) http.Handler {
 	opts := []otelhttp.Option{
```
WithTracing needs to be changed to `func WithTracing(handler http.Handler, tp trace.TracerProvider) http.Handler {` (without pointer to interface).

sallyom replied:
updated

dashpole commented on cmd/kubemark/hollow-node.go:
```diff
@@ -246,7 +246,7 @@ func run(cmd *cobra.Command, config *hollowNodeConfig) error {
 		}
 		defer fakeRemoteRuntime.Stop()
-		runtimeService, err := remote.NewRemoteRuntimeService(endpoint, 15*time.Second)
+		runtimeService, err := remote.NewRemoteRuntimeService(endpoint, 15*time.Second, nil)
```
Needs NewNoopTracerProvider()

dashpole commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,49 +26,72 @@ import (
+	resourceOpts []resource.Option,
+) (oteltrace.TracerProvider, error) {
+	if tracingConfig == nil {
+		tp := oteltrace.NewNoopTracerProvider()
+		return tp, nil
```
```suggestion
		return oteltrace.NewNoopTracerProvider(), nil
```

dashpole commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,49 +26,72 @@ import (
 	res, err := resource.New(ctx, resourceOpts...)
 	if err != nil {
-		klog.Fatalf("Failed to create resource: %v", err)
+		return nil, err
```
```suggestion
		return oteltrace.NewNoopTracerProvider(), err
```

dashpole commented on test/e2e_node/util.go:
```diff
@@ -282,7 +282,7 @@ func getCRIClient() (internalapi.RuntimeService, internalapi.ImageManagerService
 	const connectionTimeout = 2 * time.Minute
 	runtimeEndpoint := framework.TestContext.ContainerRuntimeEndpoint
-	r, err := remote.NewRemoteRuntimeService(runtimeEndpoint, connectionTimeout)
+	r, err := remote.NewRemoteRuntimeService(runtimeEndpoint, connectionTimeout, nil)
```
Needs NewNoopTracerProvider instead of nil.

sallyom replied:
updated

dashpole reviewed (commented):
Mostly suggestions to change from nil to the Noop everywhere.  Hopefully this is helpful.

sallyom commented:
will push those latest updates in a few - commented b4 pushed

sallyom commented:
/retest

liggitt commented:
Regressions and release blockers over the past two weeks starved out review time for changes like this. I'd like this but it's too late for 1.24. Will pick up review next week.

/milestone v1.25

sallyom commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,20 @@ func (l *VerbosityLevel) Set(value string) error {
+	// Endpoint of the collector this component will report traces to.
+	// The connection is insecure, and does not currently support TLS.
+	// Recommended is unset, and endpoint is the otlp grpc default, localhost:4317.
+	// +featureGate=KubeletTracing
```
updated

sallyom commented on cmd/kubelet/app/server.go:
```diff
@@ -1291,3 +1304,28 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
+			semconv.ServiceInstanceIDKey.String(componentKubelet),
+		),
+	}
+	if s.KubeletConfiguration.Tracing.Endpoint != nil {
+		opts = append(opts, otlpgrpc.WithEndpoint(*s.KubeletConfiguration.Tracing.Endpoint))
+	}
```
fixed, updated

sallyom commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -210,3 +210,20 @@ func (l *VerbosityLevel) Set(value string) error {
+	// SamplingRatePerMillion is the number of samples to collect per million spans.
+	// Recommended is unset. If unset, sampler respects its parent span's sampling
+	// rate, but otherwise never samples.
+	// +featureGate=KubeletTracing
```
updated

sallyom commented on staging/src/k8s.io/component-base/config/v1alpha1/types.go:
```diff
@@ -141,3 +141,20 @@ type VModuleItem struct {
+	// Endpoint of the collector this component will report traces to.
+	// The connection is insecure, and does not currently support TLS.
+	// Recommended is unset, and endpoint is the otlp grpc default, localhost:4317.
+	// +featureGate=KubeletTracing
```
dropped, updated

sallyom commented on staging/src/k8s.io/component-base/config/v1alpha1/types.go:
```diff
@@ -141,3 +141,20 @@ type VModuleItem struct {
+	// SamplingRatePerMillion is the number of samples to collect per million spans.
+	// Recommended is unset. If unset, sampler respects its parent span's sampling
+	// rate, but otherwise never samples.
+	// +featureGate=KubeletTracing
```
done

sallyom commented on staging/src/k8s.io/apiserver/pkg/server/options/tracing.go:
```diff
@@ -125,3 +141,24 @@ func (o *TracingOptions) Validate() (errs []error) {
+	if err := runtime.DecodeInto(codecs.UniversalDecoder(), data, internalConfig); err != nil {
+		return nil, fmt.Errorf("unable to decode tracing configuration data: %v", err)
+	}
+	tc := &componentbaseconfigv1alpha1.TracingConfiguration{
```
right, updated

sallyom commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -29,30 +29,52 @@ import (
 
-// NewProvider initializes tracing in the component, and enforces recommended tracing behavior.
-func NewProvider(ctx context.Context, baseSampler sdktrace.Sampler, resourceOpts []resource.Option, opts ...otlpgrpc.Option) trace.TracerProvider {
-	opts = append(opts, otlpgrpc.WithInsecure())
```
added back, it was in individual configureTracing fns, but is here now in InitTracing

sallyom commented:
@dashpole @liggitt  thank you again for the reviews, I believe I've updated everything based on your last reviews.

dashpole commented on cmd/kubelet/app/server.go:
```diff
@@ -1296,3 +1309,28 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
+	opts := []otlpgrpc.Option{}
+	resourceOpts := []otelsdkresource.Option{
+		otelsdkresource.WithAttributes(
+			semconv.ServiceNameKey.String(hostname),
+			semconv.ServiceInstanceIDKey.String(componentKubelet),
```
```suggestion
			semconv.ServiceNameKey.String(componentKubelet),
			semconv.HostNameKey.String(hostname),
```
host name has its own semantic convention, which we can use.  Service name [should be a constant](https://github.com/open-telemetry/opentelemetry-go/blob/8c0951ea94cb20bd9e2338b8fdc20036b8cf995b/semconv/v1.7.0/resource.go#L825).

dashpole commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
@dashpole rebased and updated the version, thanks :)

dashpole commented:
/lgtm
/approve

logicalhan commented:
/assign

sallyom commented:
> @liggitt thanks for your reviews with this PR, I believe it's ready for final approval/review

@liggitt bump, ptal, thank you again

liggitt commented:
Thanks for the ping, I ran out of review time before I'm OOO, will pick this up the week of 1.25 enhancements freeze.

sallyom commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

endocrimes commented:
/cc

endocrimes reviewed (commented):
/lgtm
for kubelet changes.

I agree with @liggitt that the embedding an optional interface is a bit risky, but I don't think I'd want to block this on it.

sallyom commented:
@endocrimes @dashpole I rebased, there are no new changes but lgtms have been wiped. Still waiting on liggitt's approval. thanks, again

liggitt commented:
Thanks for being patient. https://github.com/kubernetes/kubernetes/pull/105797 landed, which was the other in-flight PR around component-base config. Picking this back up now.

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -906,6 +916,9 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
 			utilnet.CloseIdleConnectionsFor(clientConfig.Transport)
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) && s.KubeletConfiguration.Tracing != nil {
```
It's confusing that we're checking for a non-nil `s.KubeletConfiguration.Tracing` here, but not passing it to WrapperFor. Do we need the nil check any more, or can we assume tp will be a non-nil no-op tracing provider if no tracing is configured?

halfrost replied:
I think the purpose of the if condition judgment here is that if s.KubeletConfiguration.Tracing is nil, it is meaningless to call WrapperFor(tp).

sallyom replied:
right, we don't need the nil check there, it will be the no-op. updated
(https://github.com/kubernetes/kubernetes/blob/a6544b59c7cfa6d7673b74bd02c79214b932d22d/cmd/kubelet/app/server.go#L1315-#L1331)

sallyom commented:
this last rebase was non-trivial, I'll keep rebasing as many times as is necessary, but ptal, thanks

sallyom commented:
I see a few unintended changes snuck in from the last rebase, will update to remove those

liggitt commented:
> the separation of commits has gotten wacked w/ this PR - with the co-authorship, etc.

yeah, I suspect squashing will probably be necessary

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -19,6 +19,7 @@ package config
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	componentbaseconfig "k8s.io/component-base/config"
 	logsapi "k8s.io/component-base/logs/api/v1"
```
for component-base config snippets we intend to embed inside component's config types, the logsapi was the approach we ended up with in https://github.com/kubernetes/kubernetes/pull/105797

highlights that apply to (and simplify) this PR:

1. define the type under `k8s.io/component-base/$component/api/v1` only
  * the "v1" in the path is a signal that the snippet can be used in component config files of varying stability, so we should treat the fields like a stable API (not rename them or break serialization of them)
  * since we have to treat the config serialization as stable, and don't have plans to create a v2, no "internal" version of new component-base config types is needed currently

2. in both the internal and versioned kubelet config types, use the `k8s.io/component-base/$component/api/v1` type

sallyom replied:
updated, was confused, it's clear now

liggitt commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -241,6 +242,12 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
 	}
 
+	if localFeatureGate.Enabled(features.KubeletTracing) {
+		if errs := traces.ValidateTracingConfiguration(kc.Tracing, field.NewPath("tracing")); len(errs) > 0 {
```
pass localFeatureGate into ValidateTracingConfiguration (just like logsapi.Validate), so that any future subfields of that configuration that are themselves feature-gated can guard on tracing-specific gates

sallyom replied:
added, but not using it at this time

liggitt commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -241,6 +242,12 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
+	if localFeatureGate.Enabled(features.KubeletTracing) {
+		if errs := traces.ValidateTracingConfiguration(kc.Tracing, field.NewPath("tracing")); len(errs) > 0 {
+			allErrors = append(allErrors, errs.ToAggregate().Errors()...)
+		}
```
```suggestion
		} else if kc.Tracing != nil {
			... append error about tracing configuration not being supported when the feature is disabled
		}
```

sallyom replied:
added

liggitt commented on pkg/kubelet/apis/config/validation/validation_test.go:
```diff
@@ -497,6 +498,26 @@ func TestValidateKubeletConfiguration(t *testing.T) {
 			},
 			errMsg: "invalid configuration: taint.TimeAdded is not nil",
 		},
+		{
```
add a testcase about specifying tracing configuration with the feature off returning an error

sallyom replied:
done

liggitt commented on pkg/kubelet/server/server_test.go:
```diff
@@ -348,6 +348,7 @@ func newServerTestWithDebuggingHandlers(kubeCfg *kubeletconfiginternal.KubeletCo
 		fw.fakeKubelet,
 		stats.NewResourceAnalyzer(fw.fakeKubelet, time.Minute, &record.FakeRecorder{}),
 		fw.fakeAuth,
+		nil,
```
nil or NewNoopTracerProvider?

sallyom replied:
updated to Noop

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go:
```diff
@@ -327,12 +327,12 @@ func TestEstimateMaxLengthJSON(t *testing.T) {
 						Type: "object",
```
revert changes to this file?

sallyom replied:
reverted, these changes are from the update vendor (to bring in otel pkgs) I _think_

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/value_test.go:
```diff
@@ -322,9 +322,9 @@ func TestMapValueIsSet(t *testing.T) {
 
```
revert changes to this file?

sallyom replied:
reverted, these changes are from the update vendor (to bring in otel pkgs) I _think_

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/tracing.go:
```diff
@@ -64,21 +79,21 @@ func (o *TracingOptions) ApplyTo(es *egressselector.EgressSelector, c *server.Co
 	}
 
-	errs := tracing.ValidateTracingConfiguration(npConfig)
+	errs := traces.ValidateTracingConfiguration(traceConfig, nil)
```
pass feature.DefaultFeatureGate into ValidateTracingConfiguration

sallyom replied:
done

liggitt commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -78,3 +78,18 @@ type DebuggingConfiguration struct {
 	EnableContentionProfiling bool
 }
+
+// TracingConfiguration provides versioned configuration for OpenTelemetry tracing clients.
```
drop the internal version of this type (see comment on pkg/kubelet/apis/config/types.go for an overview of where we want component-base type snippets to live)

sallyom replied:
this was a source of confusion for me, got it now, thanks

liggitt commented on staging/src/k8s.io/component-base/config/v1alpha1/types.go:
```diff
@@ -80,3 +80,18 @@ type ClientConnectionConfiguration struct {
 	Burst int32 `json:"burst"`
 }
+
+// TracingConfiguration provides versioned configuration for OpenTelemetry tracing clients.
```
move this to `staging/src/k8s.io/component-base/tracing/api/v1/types.go` (see comment on pkg/kubelet/apis/config/types.go for an overview of where we want component-base type snippets to live)

see `staging/src/k8s.io/component-base/logs/api` as an example component-base component config snippet directory structure

sallyom replied:
moved

liggitt commented on staging/src/k8s.io/component-base/traces/config.go:
```diff
@@ -14,63 +14,32 @@ See the License for the specific language governing permissions and
 */
 
-package tracing
+package traces
```
suggest `staging/src/k8s.io/component-base/tracing/api/v1`

sallyom replied:
updated

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -780,6 +781,11 @@ type KubeletConfiguration struct {
+	// See http://kep.k8s.io/2832 for more details.
+	// +featureGate=KubeletTracing
+	// +optional
+	Tracing *componentbaseconfigv1alpha1.TracingConfiguration `json:"tracing,omitempty"`
```
use "k8s.io/component-base/tracing/api/v1" (see comment on pkg/kubelet/apis/config/types.go for an overview of where we want component-base type snippets to live)

sallyom replied:
updated

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go:
```diff
@@ -150,6 +150,7 @@ type TLSConfig struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // TracingConfiguration provides versioned configuration for tracing clients.
+// This type is deprecated in favor of component-base/config TracingConfiguration
```
I'm not sure what is meant by this ... what action is expected from apiserver consumers of this config struct in response to this deprecation comment?

The component-base TracingConfiguration type doesn't have TypeMeta in it, so it's not a standalone serializable type, so consumers couldn't actually switch to using the component-base version in a file

liggitt replied:
Same question on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go

sallyom replied:
got it, removed the comments

markyjackson-taulia commented:
Hi @sallyom My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

sallyom commented:
> Hi @sallyom My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

Hi, yes, I'll be updating based on last review in the next day or 2, still on track

sallyom commented on cmd/kubelet/app/server.go:
```diff
@@ -1297,3 +1310,28 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
 	return rl, nil
 }
+
+func configureTracerProvider(s *options.KubeletServer) oteltrace.TracerProvider {
```
updated to `newTracerProvider`

sallyom commented:
@dashpole @liggitt I've implemented _all_ suggestions/feedback from the last review (due to PTO/etc, excuse the delay) thanks a ton for your guidance.
* 1 outstanding question: https://github.com/kubernetes/kubernetes/pull/105126#discussion_r915921867  I removed the gating there, ok?

liggitt commented on staging/src/k8s.io/component-base/tracing/utils.go:
```diff
@@ -26,49 +26,76 @@ import (
+	if tracingConfig == nil {
+		return oteltrace.NewNoopTracerProvider(), nil
+	}
+	opts := addedOpts
```
if we're going to mutate this later, we still need to make a copy first, e.g.:
```go
opts := append([]otlpgrpc.Option{}, addedOpts...)
```

https://go.dev/play/p/y4wXSwIxdha demonstrates how appending to a slice inside a function can modify the contents of the yet-unused capacity of the slice in ways visible to other callers of the function

sallyom replied:
updated

liggitt commented on staging/src/k8s.io/component-base/traces/utils.go:
```diff
@@ -26,49 +26,75 @@ import (
+		return oteltrace.NewNoopTracerProvider(), nil
+	}
+	if tracingConfig.Endpoint != nil {
+		opts = append(opts, otlpgrpc.WithEndpoint(*tracingConfig.Endpoint))
```
this is still an issue, see the above comment

liggitt commented on staging/src/k8s.io/component-base/tracing/api/v1/types.go:
```diff
@@ -0,0 +1,32 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
nit: 2022 for new files

sallyom replied:
updated

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/traces.go:
```diff
@@ -22,19 +22,17 @@ import (
 		otelhttp.WithPublicEndpoint(),
+		otelhttp.WithTracerProvider(tp),
 	}
-	if tp != nil {
```
in TracingOptions#ApplyTo, when TracingOptions is nil or no config file is specified, we can still be left with a nil TracerProvider instead of a no-op one, right? in that case, wouldn't we still need the nil check here and the conditional WithTracerProvider append?

sallyom replied:
I _think_ this: 
https://github.com/kubernetes/kubernetes/blob/ba7e26a2d7ad34a8d5f2d19b168a571fae869b31/staging/src/k8s.io/apiserver/pkg/server/config.go#L375
ensures TracerProvider will always be the Noop even when config is nil
so will never be nil here:
https://github.com/kubernetes/kubernetes/blob/ba7e26a2d7ad34a8d5f2d19b168a571fae869b31/staging/src/k8s.io/apiserver/pkg/server/config.go#L856 

@dashpole I'd appreciate another look at it, though, please

liggitt replied:
ok, I missed the initialization there

dashpole replied:
I'll take another pass.  As a side note, once we update to v1 of the otel libraries (blocked by etcd upgrade), passing a `nil` tracer provider will be valid, and will default to the no-op provider.

liggitt commented on staging/src/k8s.io/component-base/tracing/api/OWNERS:
```diff
@@ -0,0 +1,8 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+approvers:
+  - sig-instrumentation-approvers
```
replace this bit with this:

```
options:
  no_parent_owners: true
approvers:
  - api-approvers
```

to make sure changes route through API approval since changes here can impact compatibility of component config files

sallyom replied:
updated

liggitt reviewed (commented):
The config / API package layout / typedef / validation looks great, thanks for sticking with all the twists and turns.

I just had a few final comments on the handler and filter wiring and the API package owner file.

I didn't review this for functional testing of tracing in the kubelet, but I remember @dashpole was able to add some integration tests when he did the API server bit. There should be a test that makes sure this can actually be enabled and work, but I'll defer review of that bit to an instrumentation reviewer.

sallyom committed:
kubelet tracing: update vendor

Signed-off-by: Sally O'Malley <somalley@redhat.com>

sallyom committed:
kubelet tracing: generated files

Signed-off-by: Sally O'Malley <somalley@redhat.com>

sallyom committed:
kubelet-tracing:update

Signed-off-by: Sally O'Malley <somalley@redhat.com>

sallyom commented:
updated again! I wasn't sure about the [HandleWithFilter](https://github.com/kubernetes/kubernetes/pull/105126#discussion_r917148821) - I added the feature gate check back there, though. Also, I cleaned up the commits/squashed.

liggitt commented on pkg/kubelet/server/server.go:
```diff
@@ -126,10 +128,15 @@ type containerInterface interface {
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletTracing) {
+		handler = traces.WithTracing(handler, a.TracerProvider, "kubelet")
+	}
 	a.HandleWithFilter(path, handler)
```
> I don't understand, bc HandleWithFilter isn't called anywhere else.

There aren't any calls today, but if we add extra behavior to the Handle method, and HandleWithFilter gets called in the future, I think that would be a bug.

I finally paged back in what was going on here, and I think there were 2 ways the kubelet server was adding handlers (you can see both styles registered in `InstallDefaultHandlers()`):
1. web-service-type registration (calling `s.restfulCont.Add(ws)`)
2. raw handler path registration (calling `s.restfulCont.Handle(handler)`)

We want all the kubelet filters (before this PR, just authentication/authorization) to apply to all handlers added via both methods. That's why we intercepted calls to `Handle` and made sure they were routed to `HandleWithFilter`. Filters are already applied to the web-service style handlers.

By adding tracing inside the `Handle` override, I think that means tracing isn't getting applied to the web-service-style handlers added to the kubelet. I think a better way to add tracing to the kubelet would be via a filter which would apply to both web-service and http-style handlers, but I'm not sure how to adapt the otel handler wrapper to a restful container filter, which expects this signature:

```
func(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {
  ...
  chain.ProcessFilter(req, resp)
}
```

A test that tracing was actually working as expected on kubelet endpoints (both web-service-style and http.Handler-style) would make me a lot more confident this wiring was doing what we expected

liggitt commented:
I marked the API review bits as done, since those packages all look good.

I commented on the kubelet Handle thread as well. That wiring and a test demonstrating that is working for both styles of handler the kubelet has are the only outstanding bits I see.

dashpole reviewed (commented):
It doesn't look like the kubelet has any integration tests right now: https://github.com/kubernetes/kubernetes/tree/master/test/integration/kubelet.  I'd guess it is probably because it has so many dependencies on being on a "real" VM.  For testing the kubelet's server, we might be able to test it in [pkg/kubelet/server/server_test.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/server_test.go), and for the CRI changes, we should be able to test it in the CRI code.  I dug a bit, but don't see a good way to test context propagation generally in the kubelet without a full node e2e test

Here is the APIServer integration test: https://github.com/kubernetes/kubernetes/blob/master/test/integration/apiserver/tracing/tracing_test.go

liggitt commented on staging/src/k8s.io/component-base/tracing/api/v1/config.go:
```diff
@@ -14,63 +14,32 @@ See the License for the specific language governing permissions and
-
 // ValidateTracingConfiguration validates the tracing configuration
-func ValidateTracingConfiguration(config *apiserver.TracingConfiguration) field.ErrorList {
+func ValidateTracingConfiguration(traceConfig *TracingConfiguration, featureGate featuregate.FeatureGate, fldPath *field.Path) field.ErrorList {
```
I asked for it to be plumbed for parity with the log config validation... if specific fields are added that are at a different stability level, this would use feature enablement to determine whether it was valid for them to be set

sallyom committed:
add otelrestful restful.FilterFunction

Signed-off-by: Sally O'Malley <somalley@redhat.com>

sallyom committed:
add runtime-service test with tracerProvider

Signed-off-by: Sally O'Malley <somalley@redhat.com>

sallyom commented:
> It doesn't look like the kubelet has any integration tests right now: 
https://github.com/kubernetes/kubernetes/tree/master/test/integration/kubelet. I'd guess it is probably because it has so many dependencies on being on a "real" VM

I added a remote-service with tracer provider test, also, I've confirmed with local testing that spans are generated. I can add a more in-depth test/integration test as a follow-up. 

![kubelet-crio-span](https://user-images.githubusercontent.com/11166065/180339739-57e8683b-b6c4-4ad9-bd67-56c1f821fe91.png)
![spans](https://user-images.githubusercontent.com/11166065/180339917-be5d2c76-cff2-43db-bfef-79d862629d6e.png)

sallyom commented:
/test pull-kubernetes-e2e-kind

sallyom commented:
for explanation on how to run K8s (kubeadm) with etcd, crio, apiserver, & kubelet tracing enabled, see this:
https://github.com/sallyom/otel-kubeadm/blob/main/README.md#launch-kubeadm-cluster

markyjackson-taulia commented:
Hello @sallyom, my name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Is this still on track for 1.25?

sallyom commented:
> 

AFAIK, yes. API review is complete. @dashpole anything else or is this ready to merge?

dashpole commented on pkg/kubelet/cri/remote/remote_runtime_test.go:
```diff
@@ -55,6 +60,18 @@ func createRemoteRuntimeService(endpoint string, t *testing.T) internalapi.Runti
+	ep := "localhost:4317"
+	sr := int32(99999)
+	traceConfig := &tracingapi.TracingConfiguration{Endpoint: &ep, SamplingRatePerMillion: &sr}
+	tp, err := tracing.NewProvider(context.Background(), traceConfig, []otlpgrpc.Option{}, []otelsdkresource.Option{})
```
This is good for testing the config-providing code, but not as helpful for testing that requests generate spans.  Could we instead use the [InMemoryExporter](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/tracetest/exporter.go#L47) to check that it is generating a span for the version call?

https://github.com/open-telemetry/opentelemetry-go/blob/1eae91b3b099af62100cd630cd1cab89fe841615/sdk/trace/batch_span_processor_test.go#L575 is an example test you can model it after.

sallyom replied:
added a new test to verify spans are exported with version call

sallyom commented:
/test pull-kubernetes-e2e-gce-alpha-features

dashpole commented on simple-pod.yaml:
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
```
You probably didn't mean to commit this file.

sallyom replied:
right i did not

sallyom replied:
updated, removed the file, thanks

dashpole commented:
LGTM otherwise

endocrimes reviewed (commented):
/approve
^ kubelet e2e change

LGTM from a node perspective. I'd like us to work on an E2E post-merge, but wouldn't want to block this from landing on figuring one out.

dashpole commented:
/lgtm
/approve

sallyom commented:
@dashpole @endocrimes  thanks again for your reviews, i think this one is ready to go w/ your final approvals :partying_face:

sallyom commented:
/retest

derekwaynecarr reviewed (commented):
I am not an expert in tracing so apologies for naive questions.

I can see in kubelet code paths that we intercept all remote runtime interactions and we install a filter that intercepts all HTTP operations served by the kubelet itself, so practically the configuration and intercepts all seem to be at the proper place.

What I don't see is anything in an integration test (or elsewhere) that looks at the emitted traces and validates them in any way?  Is there anything that I may have missed that checks we are actually tracing these things?

liggitt commented:
(fwiw, API and dependency updates still lgtm, let me know when the kubelet package has approval and I can tag)

dashpole commented:
> Is there anything that I may have missed that checks we are actually tracing these things?

https://github.com/kubernetes/kubernetes/pull/105126/files#diff-1315cc53cf9b0369fb7c9ef40b24e8e8c136ad0c8f3c75db4d3c12a6263966d3R73 is a unit test (using an in-memory span exporter) to test that a CRI call generates a span.  We could add similar tests for the kubelets' http servers as well.

An alternative would be to add a node e2e test that actually runs a collector and validates the output.

dashpole commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -77,10 +81,23 @@ func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration) (
+		}
+		// Even if there is no TracerProvider, the otelgrpc still handles context propagation.
+		// See https://github.com/open-telemetry/opentelemetry-go/tree/main/example/passthrough
+		dialOpts = append(dialOpts,
```
This currently is just a generic span for kubelet -> cri interactions.  Ideally, we would add a single parent for a syncPod loop which has that information and also ties together multiple CRI calls (e.g. start pod sandbox, pull image, create container, etc).

derekwaynecarr commented:
thanks @dashpole 

/approve
/lgtm

liggitt commented:
/approve
for API and dep update

githubsands commented:
LGTM

halfrost commented on pkg/kubeapiserver/admission/config.go:
```diff
@@ -47,7 +47,7 @@ type Config struct {
 
 // New sets up the plugins and admission start hooks needed for admission
-func (c *Config) New(proxyTransport *http.Transport, egressSelector *egressselector.EgressSelector, serviceResolver webhook.ServiceResolver, tp *trace.TracerProvider) ([]admission.PluginInitializer, genericapiserver.PostStartHookFunc, error) {
+func (c *Config) New(proxyTransport *http.Transport, egressSelector *egressselector.EgressSelector, serviceResolver webhook.ServiceResolver, tp trace.TracerProvider) ([]admission.PluginInitializer, genericapiserver.PostStartHookFunc, error) {
```
Just a small quick question. Why doesn't the input trace.TracerProvider parameters here use pointer types? I am working on a similar feature. I happened to see you modify here. So wanted to ask your idea.

Aneurysm9 replied:
`trace.TracerProvider` [is an interface](https://pkg.go.dev/go.opentelemetry.io/otel/trace#TracerProvider).  I think it would typically be most appropriate to receive the interface and not a pointer to the interface.

sallyom replied:
*I think this is the answer but anyone can correct me here or make it clearer, please*
During reviews it was decided to modify from the pointer parameter, as `pointers to interfaces` are very rarely required and are discouraged. ([Here's a good explanation I learned from](https://go.dev/doc/faq#pointer_to_interface)) Instead, methods that satisfy interfaces usually have pointer receivers. This is because if the method needs to modify the receiver, the receiver must be a pointer, because changes to that receiver will be invisible to the caller. Here's another good post about how [interfaces are pointers themselves](https://www.timr.co/go-interfaces-the-tricky-parts/#:~:text=Go's%20interface%20values%20are%20really,type%20of%20the%20concrete%20value.) (and a pointer to an interface would be a double pointer).

Here's the [TracerProvider interface](https://github.com/kubernetes/kubernetes/blob/d0c92aedbab1bd448a5271496689ae2a3f11debd/vendor/go.opentelemetry.io/otel/trace/trace.go#L663) - the parameter highlighted above
Here's the [TracerProvider struct](https://github.com/kubernetes/kubernetes/blob/eefcf6aa801c5db48b564d0464470d623b8bfb79/vendor/go.opentelemetry.io/otel/sdk/trace/provider.go#L109)
You can see the go sdk also does not use pointer to the TracerProvider as param, like here [WithTracerProvider](https://github.com/kubernetes/kubernetes/blob/d0c92aedbab1bd448a5271496689ae2a3f11debd/vendor/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/config.go#L86)

halfrost replied:
@sallyom Thanks sallyom. The reason you gave makes so much sense! This pr is perfect.
